### DISCUSSION TITLE
Questions about Robot Inventor/SPIKE Prime hub (gestures and bluetooth)

### DISCUSSION THREAD

**Discussion by Mantarri** [pybricks/support#1624] [Q&A]
I got a Robot Inventor set last week and am starting to put it through its paces with PyBricks after some initial testing with the stock firmware and I have a few questions (also, for clarity, I'm using PyBricks' MicroPython, not block coding.)

1. How would I detect gestures? The official firmware had some functions for checking if the hub was tapped, falling, etc. I don't see anything pertaining to that explicitly in the PyBricks docs, so I assume you'd do so off the raw accelerometer readings, but I'm not sure what the "proper" implementation approach would be.
2. What would it take to be able to interact with bluetooth classic devices? Most of the Powered Up hubs only support BLE, but the Robot Inventor/SPIKE Prime hubs do support BT Classic from my understanding. I'm the same person from #1452 and am willing to also take a stab at this if I can figure it out. I was thinking perhaps in the form of two classes; generic interfaces for Bluetooth Classic and BLE respectively which users could then script on top of to send messages/write to command characteristics/listen/etc? I could fold my BuWizz communication into that as was suggested by just then providing an example script for the BLE class. Perhaps an NXT brick for an example for the BT Classic class (since I have one, and I've seen some interest in using NXT bricks with Python.) The BT Classic communication  module would of course only be for the Mindstorms/SPIKE Prime hubs (and theoretically the SPIKE essential hub as well from my understanding. I just don't have one.)
3. What's a nice, clean approach to using one PyBricks hub as a remote for another using a sort of event-based system? I.e. I press a button on the inventor hub and my technic hub responds to that event? I've messed around with this, but got stuck. I was trying to make sure BLE broadcasts were received, but my approach of continuing a broadcast until the other hub acknowledged it ran into problems as then the target hub would run the command multiple times due to an acknowledgement taking too long and allowing the command to be sent multiple times. I tried to look at the example projects for using PyBricks with the train remote as an example, but it appears that the remote class abstracts away the problem of ensuring commands are received. I'm guessing that's because the remote connects to the hub, whereas the broadcasts are just... broadcasts? And so there isn't any low-level code that establishes a reliable connection?

**Comment by laurensvalk** [comment]
> How would I detect gestures? 

If you'd like to detect free-fall, you could check the magnitude of the acceleration vector, which should be small in that case.

Tapping could be detected similarly. You might even be able to get a crude estimation of the direction of the tap.

The builtin gesture filters you are referring to are not enabled. If there is a strong use case for them, we might try to do that in the future.

> What would it take to be able to interact with bluetooth classic devices?

I suppose you could look into whether and how the BTStack library that we use for Bluetooth on the SPIKE/Mindstorms hubs supports combining BLE and classic. We haven't really needed it so far.

> What's a nice, clean approach to using one PyBricks hub as a remote for another using a sort of event-based system?

Broadcasting is good for communicating states, but less so for events. Depending on the application, it often works best to use one hub as the main controller, and use the observing hub as an external source of motors and sensors.

So you could maintain button press states on the main hub, and communicate the action to the observing hub instead of the event.

-----------

For example, say you had a leading hub with a button and you want to have the follower hub rotate a motor +90 degrees every time you press the button.

Instead of communicating the button press event, you could broadcast the target state of 90, 180, 270, incrementing on every button press. This is persistent over time, and acknowledgement may not even be needed.

**Reply by Mantarri** [reply]
> If you'd like to detect free-fall, you could check the magnitude of the acceleration vector, which should be small in that case.
> 
> Tapping could be detected similarly. You might even be able to get a crude estimation of the direction of the tap.
> 
> The builtin gesture filters you are referring to are not enabled. If there is a strong use case for them, we might try to do that in the future.

Thanks. Will do some experimenting with that then. Also, for curiosity's sake, are the built-in filters essentially just pre-made "detection profiles," or is there more to them?

> I suppose you could look into whether and how the BTStack library that we use for Bluetooth on the SPIKE/Mindstorms hubs supports combining BLE and classic. We haven't really needed it so far.

Looking at BTStack's documentation, it does indeed support Classic in addition to LE, including dual mode according to one use example. Their documentation looks good at a glance, so I'll further pursue this there unless anything PyBricks-specific comes up as for as functionality goes. Besides that, thoughts on my theoretical implementation?
> Broadcasting is good for communicating states, but less so for events. Depending on the application, it often works best to use one hub as the main controller, and use the observing hub as an external source of motors and sensors.
> 
> So you could maintain button press states on the main hub, and communicate the action to the observing hub instead of the event.

Ah, gotcha. So I suppose my theoretical generic BLE client class would be the way to go for an event approach; have the controller hub use it to connect to the output hub and send commands via writing to its stdin.
